04-onion-routing.md

```diff
@@ -435,85 +426,91 @@ the obfuscated `hops_data`.
+	filler := generateFiller("rho", path, hopSharedSecrets)
+	// Allocate zero'd out byte slices to store the final mix header packet
+	// and the hmac for each hop.
+	var (
+		mixHeader  [routingInfoSize]byte
+		nextHmac   [hmacSize]byte
+		hopDataBuf bytes.Buffer
+	)
+
+	// Now we compute the routing information for each hop, along with a
+	// MAC of the routing info using the shared key for that hop.
 	for i := numHops - 1; i >= 0; i-- {
-		rhoKey := generateKey("rho", hopSharedSecrets[i])
-		muKey := generateKey("mu", hopSharedSecrets[i])
-
-		hopsData[i].HMAC = nextHmac
+		// We'll derive the two keys we need for each hop in order to:
+		// generate our stream cipher bytes for the mixHeader, and
+		// calculate the MAC over the entire constructed packet.
+		rhoKey := generateKey("rho", &hopSharedSecrets[i])
+		muKey := generateKey("mu", &hopSharedSecrets[i])
+
+		// The HMAC for the final hop is simply zeroes. This allows the
+		// last hop to recognize that it is the destination for a
+		// particular payment.
+		path[i].HopPayload.HMAC = nextHmac
+
+		// Next, using the key dedicated for our stream cipher, we'll
+		// generate enough bytes to obfuscate this layer of the onion
+		// packet.
+		streamBytes := generateCipherStream(rhoKey, routingInfoSize)
+
+		// Before we assemble the packet, we'll shift the current
+		// mix-header to the right in order to make room for this next
+		// per-hop data.
+		rightShift(mixHeader[:], path[i].HopPayload.CountFrames()*frameSize)
+
+		// With the mix header right-shifted, we'll encode the current
+		// hop data into a buffer we'll re-use during the packet
+		// construction.
+		if err := path[i].HopPayload.Encode(&hopDataBuf); err != nil {
+			return nil, err
+		}
 
-		// Shift and obfuscate routing information
-		streamBytes := generateCipherStream(rhoKey, numStreamBytes)
+		copy(mixHeader[:], hopDataBuf.Bytes())
 
-		rightShift(mixHeader[:], hopDataSize)
-		buf := &bytes.Buffer{}
-		hopsData[i].Encode(buf)
-		copy(mixHeader[:], buf.Bytes())
-		xor(mixHeader[:], mixHeader[:], streamBytes[:routingInfoSize])
+		// Once the packet for this hop has been assembled, we'll
+		// re-encrypt the packet by XOR'ing with a stream of bytes
+		// generated using our shared secret.
+		xor(mixHeader[:], mixHeader[:], streamBytes[:])
 
-		// These need to be overwritten, so every node generates a correct padding
+		// If this is the "last" hop, then we'll override the tail of
+		// the hop data.
 		if i == numHops-1 {
 			copy(mixHeader[len(mixHeader)-len(filler):], filler)
 		}
 
+		// The packet for this hop consists of: mixHeader. When
+		// calculating the MAC, we'll also include the optional
+		// associated data which can allow higher level applications to
+		// prevent replay attacks.
 		packet := append(mixHeader[:], assocData...)
 		nextHmac = calcMac(muKey, packet)
+
+		hopDataBuf.Reset()
 	}
 
-	packet := &OnionPacket{
-		Version:      0x00,
+	return &OnionPacket{
+		Version:      baseVersion,
```

Comment:
Don't see `baseVersion` defined anywhere.

Reply:
Right, I copied it from the code. Will set it back to `0x00`.